resources/fetching/fetcher.go

```diff
@@ -31,7 +32,7 @@ type Factory interface {
 
 // Fetcher represents a data fetcher.
 type Fetcher interface {
-	Fetch(context.Context) ([]Resource, error)
+	Fetch(context.Context, chan<- ResourceInfo, CycleMetadata) error
```

Comment:
Honestly, I am not in favor of doing that change before we have more information about the requirement.
@oren-zohar is about to start working on an epic called "Cloudbeat performance" that will give us more knowledge on this.

Disadvantages of doing this premature optimization:
1. Changing a clear and simple API to a more complicated one.
2. Working with channels could make some trouble that we can't expect.
3. Right now, the channel size is 0, meaning every time we push some object into it, we can't push another one until the next step of the pipeline is pulling it. It will cause some kind of ping-pong of context switches between the sender and the receiver.

I think we can avoid the `fetcher` change on this PR, and deal with it in the future only if we understand it's inevitable.

Reply:
Thanks, @amirbenun.
This is the first step of the pipeline, therefore it determines the pace for the rest of the steps.
Moreover, due to this addition, we don't need to wait for the manager synchronization- once a resource is collected it's pushed to the next step for processing.

> Right now, the channel size is 0, meaning every time we push some object into it, we can't push another one until the next step of the pipeline is pulling it. It will cause some kind of ping-pong of context switches between the sender and the receiver.

I agree, it might be better to use a buffered channel in order to not be blocked.

Reply:
Thanks for the reply @uri-weisman 
> Moreover, due to this addition, we don't need to wait for the manager synchronization- once a resource is collected it's pushed to the next step for processing.

I didn't say the manager should wait for **all fetchers** to finish. All I am saying is the manager should wait for **a fetcher** to complete a cycle before it publishes its resources to the next pipeline's steps. Other fetchers won't be affected by a corrupted one.
It will keep a clear API and may save some trouble in the future.

Reply:
Since your comment I made a minor change in which the fetcher receives the `resource ch` upon creation and not as an argument, nevertheless, channels are still in use.

In the future, I guess we don't want to collect all resources and only then start processing (CSPM use case), Eyal commented about it as well - https://github.com/elastic/cloudbeat/pull/201#pullrequestreview-991871121

I understand your point regarding channels, and most definitely I'll analyze how system resources ( memory, cpu, etc ) have changed due to this PR.

Reply:
Sounds good, anyway I think the process should be the opposite.
We should first have a benchmark of how cloudbeat affects the system, and only then we should think of how it should be improved. Pushing such a change must be backed up with numbers.